doc/proposals/webhook-proposal.md

```diff
@@ -0,0 +1,137 @@
+```go
+type webhookServer struct {
+    Port int
+    CertDir string
+} 
+```
+
+The initialization function will be called in the main.go file, regardless of whether or not a user opts in, however no actions are taken unless the user chooses to generate webhooks.
+
+```go
+func WebhookInit() error {
+    log.Info("Starting webhook server")
+    hookServer := &webhook.Server{
+        Port: hookServerCfg.Port,
+        CertDir: hookServerCfg.CertDir,
+    }
+    
+    if err := mgr.Add(hookServer); err != nil {
+        log.Error(err, "Unable to register webhook server")
+        return err
+    }
+
+    log.Info("Registering webhooks to the webhook server")
+    
+    hookServer.Register("/mutate-pods", &webhook.Admission{Handler: &podAnnotator{}})
+    hookServer.Register("/validate-pods", &webhook.Admission{Handler: &podValidator{}})
+    
+    return nil
+}
+```
+
+
+In addition to the above file being used to start up the server, an additional  files (`mutationswebhook.go` and `validationwebhooks.go`) will generated when the user chooses to include webhooks. These are the files which have the necesary boilerplate for a user to implement the `Handle` function in order to run custom logic on validation and admission of pods.
+
+```go
+type podInteraction struct {
+	client  client.Client
+	decoder *admission.Decoder
+}
+
+func (i *podInteraction) Handle(ctx context.Context, req admission.Request) admission.Response {
+	pod := &corev1.Pod{}
+
+	err := i.decoder.Decode(req, pod)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+    //Insert mutation/validation logic here
+
+    //BRANCH
+    //Block A contains the boiler plate logic for patching a mutation
+    //Block B contains the boilder plate logic for 
+
+    //BLOCK A
+	marshaledPod, err := json.Marshal(pod)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+    return admission.PatchResponseFromRaw(req.Object.Raw, marshaledPod)
+    
+    //BLOCK B
+    if FAILURE_CONDITION {
+		return admission.Denied(fmt.Sprintf("reason for failure"))
+	}
+
+    return admission.Allowed("")
+    
+    //END BRANCH
+}
+
+// podInteraction implements inject.Client.
+// A client will be automatically injected.
+
+// InjectClient injects the client.
+func (i *podInteraction) InjectClient(c client.Client) error {
+	i.client = c
+	return nil
+}
+
+// podInteraction implements admission.DecoderInjector.
+// A decoder will be automatically injected.
+
+// InjectDecoder injects the decoder.
+func (i *podInteraction) InjectDecoder(d *admission.Decoder) error {
+	i.decoder = d
+	return nil
+}
+
+```
+
+
+### User facing usage (if needed)
+
+My suggested method for interacting with this feature is to have a command in the osdk which can be run after generating the base operator. The new command `$ operator-sdk generate webhook` will write the files cmd/manager/osdk-webhook.go, cmd/manager/mutationwebhook.go, and cmd/manager/validationwebhook.go
```

Comment:
Could we have these files live in the users webhook package. e.g. `pkg/webhook/validationwebhook.go` and so on. This will make it easier to migrate to kubebuilder as they have only one `main.go` file that lives in the [root of the project.](https://github.com/kubernetes-sigs/kubebuilder/tree/ba0d56052ea77afd0d7fa0a9d2a3b923cdadb307/testdata/project-v2)